Fix #851: Show image title when long-pressing an image

cndouglas opened the pull request:
Fix #851

This change sets the context menu's title to the image's title (if it exists). Otherwise, it shows the image's URL, which is the existing behavior.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.


## Test Plan:

1. Go to https://xkcd.com/2107/
2. Long-press on the comic image.
3. The image's title should be displayed above the Open Image In New Tab button, as shown in the screenshot below.

### Screenshots:

See https://github.com/brave/brave-ios/pull/875#issuecomment-483455131

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [ ] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

iccub commented on Client/Assets/AllFramesAtDocumentEnd.js:
```diff
@@ -1 +1 @@
-!function(e){var n={};function t(r){if(n[r])return n[r].exports;var o=n[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,t),o.l=!0,o.exports}t.m=e,t.c=n,t.d=function(e,n,r){t.o(e,n)||Object.defineProperty(e,n,{configurable:!1,enumerable:!0,get:r})},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return ...
\ No newline at end of file
+!function(e){var n={};function t(r){if(n[r])return n[r].exports;var o=n[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,t),o.l=!0,o.exports}t.m=e,t.c=n,t.d=function(e,n,r){t.o(e,n)||Object.defineProperty(e,n,{configurable:!1,enumerable:!0,get:r})},t.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(n,"a",n),n},t.o=function(e,n){return ...
```
I think you have to remove this change from the PR, otherwise our CI throws an error.
This line sometimes get changed when running the bootstrap script, sorry about that

cndouglas replied:
I forgot to include `ContextMenu.js` in the commit. Fixed.

jhreis commented on Client/Frontend/Browser/ContextMenuHelper.swift:
```diff
@@ -91,7 +92,9 @@ extension ContextMenuHelper: TabContentScript {
 
         if linkURL != nil || imageURL != nil {
-            elements = Elements(link: linkURL, image: imageURL)
+            let rawTitle = data["title"] as? String ?? ""
```
For consistency sake, I think following the pattern that `image` and `link` use would probably be best:

```
var titleText: String?
if let title = data["title"] as? String, !title.isEmpty {
    titleText = title
}
```

...

```
if linkURL != nil || imageURL != nil {
    elements = Elements(link: linkURL, image: imageURL, title: title)
}

cndouglas replied:
Fixed.

jhreis reviewed (changes requested):
Still waiting on a security review, but did leave one comment. However, feel free to wait for sec-review before making changes. Don't want to waste your time 😄

cndouglas commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -2485,7 +2485,7 @@ extension BrowserViewController: ContextMenuHelperDelegate {
         if let url = elements.image {
             if dialogTitle == nil {
-                dialogTitle = url.absoluteString
+                dialogTitle = elements.title ?? url.absoluteString.ellipsize(maxLength: ActionSheetTitleMaxLength)
```
I moved the `ellipsize` method to here (from line 2562) so that only URLs are "ellipsized". Titles will always be shown in full, which matches Safari's behavior.

jhreis replied:
Does this put an ellipse'd URL solely on one single line?

cndouglas replied:
From a quick test, looks like the URL may be up to around 3 lines long.
![](https://user-images.githubusercontent.com/19424103/55576022-df8b9380-56d5-11e9-95f7-80b0805c5a17.png)

jhreis reviewed (changes requested):
@cndouglas Thanks for your patience here. So talked with both sec and design, and they want both
- [ ] title
- [ ] full url

Potentially will be a lot of information in here. Give `title\nurl` and `url\ntitle`, and see which one feels more natural.

Additionally, security is more leaning towards full URL, so I think we'll go with that for now.

cndouglas commented:
@jhreis
I removed the "ellipsize" call, so now the title and URL are always shown in full.

I went with the Title-URL option but made it easy to switch if desired.

I also made the current URL behavior behavior more explicit in the code. As I noted on line 2491, "If the image is a link, show the link's URL. Otherwise, show the image's source URL."

Here are examples of what the two options look like (for the image on https://xkcd.com/2131/):
Title-URL (this PR):
![TitleURL](https://user-images.githubusercontent.com/19424103/55577727-d7355780-56d9-11e9-9a03-a9d8d9779ec8.png)
URL-Title (alternative):
![URLTitle](https://user-images.githubusercontent.com/19424103/55577728-d7cdee00-56d9-11e9-9f11-10e923e8f868.png)

cndouglas commented:
@jhreis I just sent you an email.

jhreis commented:
@cndouglas thanks! Just running this by our sec team again for final approval.

jhreis commented:
@cndouglas looks super great. Sounds like the consensus is:

```
title\n 
\n
url
```

e.g.:

> This is an amazing title
>
> someurl.tld

cndouglas commented:
Per https://github.com/brave/brave-ios/pull/875#issuecomment-482181660, I added another space between the title and the URL, so now there are two spaces (`\n\n`). Here's what it looks like now:
![](https://user-images.githubusercontent.com/19424103/55984001-13306580-5c62-11e9-9bd6-cd9d104b9ebd.png)

jhreis commented:
@cndouglas _perfect_ Thanks! I am submitting a security review request, so will merge it once that has been completed. Thanks!

jumde commented:
![image](https://user-images.githubusercontent.com/1903815/56157125-96f49580-5f73-11e9-8826-d20abdb27b77.png)

Text overflow hides the URL. We should have text limits to avoid this case.

cndouglas commented:
We could do a character limit for the title. However, the website could theoretically get around the limit by using dozens of newlines in the title.

iccub commented:
We could trim whitespace characters, right?

jhreis commented:
Alright, sorry for the insane changes, but after sec review, we decided to switch the title / url. If the text overflows, we want to guarantee that the user can at least see the full URL.

cndouglas committed:
Fix #851: Show image title when long-pressing an image

cndouglas commented:
I switched the URL and title. I also fixed the logic if the title is not nil but empty (line 2483).

Screenshot:
![](https://user-images.githubusercontent.com/19424103/56171688-cf17ca80-5fab-11e9-916d-f055d3a40ff9.png)

jhreis commented:
This is great, thanks!
